device-driver/src/register.rs

```diff
@@ -2,74 +2,150 @@ use bitvec::{array::BitArray, field::BitField};
-    fn read_register<R, const SIZE_BYTES: usize>(
+    fn read_register<const SIZE_BYTES: usize>(
         &mut self,
+        address: Self::AddressType,
         data: &mut BitArray<[u8; SIZE_BYTES]>,
-    ) -> Result<(), Self::Error>
-    where
-        R: Register<SIZE_BYTES, AddressType = Self::AddressType>;
+    ) -> Result<(), Self::Error>;
 }
 
 /// A trait to represent the interface to the device.
 ///
 /// This is called to asynchronously write to and read from registers.
-pub trait AsyncRegisterDevice {
+pub trait AsyncRegisterDevice: AddressableDevice {
     /// The error type
     type Error;
-    /// The address type used by this interface. Should likely be an integer.
-    type AddressType;
 
     /// Write the given data to the register asynchronously.
     ///
     /// The address and the length of the register is descriped with the constant in the `R` [Register] type.
     /// The data can made into a normal slice by calling `as_raw_slice` on it.
-    async fn write_register<R, const SIZE_BYTES: usize>(
+    async fn write_register<const SIZE_BYTES: usize>(
         &mut self,
+        address: Self::AddressType,
         data: &BitArray<[u8; SIZE_BYTES]>,
-    ) -> Result<(), Self::Error>
-    where
-        R: Register<SIZE_BYTES, AddressType = Self::AddressType>;
+    ) -> Result<(), Self::Error>;
 
     /// Read the data from the register asynchronously into the given buffer.
     ///
     /// The address and the length of the register is descriped with the constant in the `R` [Register] type.
     /// The data can made into a normal slice by calling `as_raw_slice` on it.
-    async fn read_register<R, const SIZE_BYTES: usize>(
+    async fn read_register<const SIZE_BYTES: usize>(
         &mut self,
+        address: Self::AddressType,
         data: &mut BitArray<[u8; SIZE_BYTES]>,
-    ) -> Result<(), Self::Error>
-    where
-        R: Register<SIZE_BYTES, AddressType = Self::AddressType>;
+    ) -> Result<(), Self::Error>;
+}
+
+/// A trait to represent a block of registers.
+pub trait RegisterBlock {
+    /// The type of device this register block belongs to.
+    type Device: AddressableDevice;
+
+    /// The base address to use for registers in this register block.
+    ///
+    /// This base address will be added to the register's address when reading or writing the register to the parent device.
+    const BASE_ADDR: <Self::Device as AddressableDevice>::AddressType;
+
+    /// The device this register block belongs to.
+    fn dev(&mut self) -> &mut Self::Device;
+}
+
+impl<T: RegisterBlock> AddressableDevice for T {
+    type AddressType = <T::Device as AddressableDevice>::AddressType;
+}
+
+impl<T: RegisterBlock> RegisterDevice for T
+where
+    T::Device: RegisterDevice,
+    <T::Device as AddressableDevice>::AddressType: Add<
+        <T::Device as AddressableDevice>::AddressType,
+        Output = <T::Device as AddressableDevice>::AddressType,
+    >,
+{
```

Comment:
This feels weird. Implement the device traits on a register block? Feels kinda inverted

Reply:
This allows you to define a `RegisterBlock` just like you define a device. So you can use the `implement_device!` macro on a struct implementing `RegisterBlock`.

It also allows for `RegisterBlock`s to be composed. So you could have (if you wanted) a `RegisterBlock` that contains other `RegisterBlock`s, etc.